### PR TITLE
Handle cancellation in `EventsWorkerStore._get_events_from_cache_or_db`

### PR DESCRIPTION
**Pull request by squahtx**
Multiple calls to `EventsWorkerStore._get_events_from_cache_or_db` can
reuse the same database fetch, which is initiated by the first call.
Ensure that cancelling the first call doesn't cancel the other calls
sharing the same database fetch.

Signed-off-by: Sean Quah <seanq@element.io>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/storage/databases/main/events_worker.py [hunk 1]
```diff
@@ -640,42 +640,57 @@ async def _get_events_from_cache_or_db(
...
+                        missing_events_ids,
+                    )
+                except Exception as e:
+                    with PreserveLoggingContext():
+                        fetching_deferred.errback(e)
+                    raise e
+                finally:
+                    # Ensure that we mark these events as no longer being fetched.
+                    for event_id in missing_events_ids:
+                        self._current_event_fetches.pop(event_id, None)
```
Comment (DMRobertson) [inline synapse/storage/databases/main/events_worker.py:680]: Why `None` here? Would we ever expect some other async bit of execution to remove this event from `_current_event_fetches`?

File: synapse/storage/databases/main/events_worker.py [hunk 2]
```diff
@@ -640,42 +640,57 @@ async def _get_events_from_cache_or_db(
...
+                # going to pull from the DB. We use a single deferred that resolves
+                # to all the events we pulled from the DB (this will result in this
+                # function returning more events than requested, but that can happen
+                # already due to `_get_events_from_db`).
+                fetching_deferred: ObservableDeferred[
+                    Dict[str, EventCacheEntry]
+                ] = ObservableDeferred(defer.Deferred(), consumeErrors=True)
                 for event_id in missing_events_ids:
-                    self._current_event_fetches.pop(event_id, None)
+                    self._current_event_fetches[event_id] = fetching_deferred
```
Comment (DMRobertson) [inline synapse/storage/databases/main/events_worker.py:662]: What happens if `event_id` already exists in `self._current_event_fetches`? Perhaps this is never the case because of how `missing_events_ids` is constructed, line 640 in particular?

File: tests/storage/databases/main/test_events_worker.py [hunk 3]
```diff
@@ -281,3 +282,119 @@ def test_recovery(self) -> None:
...
+
+            # Unblock the database fetch.
+            unblock.callback(None)
+            # A `CancelledError` should be raised out of the first `get_event` call.
+            exc = self.get_failure(get_event1, CancelledError).value
+            self.assertIsInstance(exc, CancelledError)
+            # The second `get_event` call should complete successfully.
+            self.get_success(get_event2)
+
+    def test_second_get_event_cancelled(self):
```
Comment (DMRobertson) [inline tests/storage/databases/main/test_events_worker.py:386]: What happens/should happen if both `get_event`s are cancelled? Should the database query be automatically cancelled somehow?

**Review by DMRobertson — COMMENTED** [review]
Seems reasonable to me. What's the benefit of this change---prep work for having cancelled requests propagate through to cancelling the underlying queries?

File: synapse/storage/databases/main/events_worker.py (hunk 2 above)
Comment (squahtx) [reply synapse/storage/databases/main/events_worker.py:662]: I don't think that can happen because of how `missing_event_ids` is constructed: `already_fetching_ids` is `intersection(missing_event_ids, _current_event_fetches)`.
If it did manage to happen, it'd mean we're doing extra database fetches that could be avoided.

Note that all this is just the previous code, indented into a function.

File: synapse/storage/databases/main/events_worker.py (hunk 1 above)
Comment (squahtx) [reply synapse/storage/databases/main/events_worker.py:680]: I have no idea. It's from the previous version of the code, introduced in #10119.

`event_id` would only be missing if we managed to replace an existing fetch above, which we shouldn't do.

File: tests/storage/databases/main/test_events_worker.py (hunk 3 above)
Comment (squahtx) [reply tests/storage/databases/main/test_events_worker.py:386]: That's a good question. Currently the database query will complete with nothing waiting for it. This is mainly for simplicity's sake. We've been following the same sort of pattern in other places where there are operations that are shared or risky to cancel.

Letting the cancellation go deeper is a lot more work. There is a pool of up to 3 event-fetcher threads which service queues of event ids. Each thread uses the same database connection for its entire lifetime. We'd need:
* a way to determine whether there's anything waiting for `get_missing_events_from_db`'s fetch
* logic to dequeue or skip cancelled items in the queue
* and optionally a way to cancel an in-flight db query without wrecking the event fetcher thread
tl;dr. improving the granularity of cancellations is a lot of extra work that I'm not yet convinced is worth it.

**Comment by squahtx** [comment]
> What's the benefit of this change---prep work for having cancelled requests propagate through to cancelling the underlying queries?

It's to prevent the cancellation of a request from leaking over to a different unrelated request that happens to be retrieving the same events. ie. the first test case fails without this change. There's no plan to do any cancellation at the db query level right now.

File: synapse/storage/databases/main/events_worker.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/events_worker.py:680]: > event_id would only be missing if we managed to replace an existing fetch above, which we shouldn't do.

That was my suspicion---thanks for confirming!

**Comment by DMRobertson** [comment]
> It's to prevent the cancellation of a request from leaking over to a different unrelated request that happens to be retrieving the same events. ie. the first test case fails without this change. There's no plan to do any cancellation at the db query level right now.

Thanks, that's very clear. 

>  ie. the first test case fails without this change

Does this mean this change is a `.bugfix` rather than a `.misc` change?

**Comment by squahtx** [comment]
> Does this mean this change is a .bugfix rather than a .misc change?

I think it's more of a feature than a bugfix. We're adding cancellation "support" to a place where there was previously none.
It depends on whether the intent of the previous code was to support cancellation, I guess?

**Comment by squahtx** [comment]
Thanks for the review!
